File: ios/Ubapp/Games/Tag/TagTransport.swift
import Foundation

/// Bidirectional channel for tag messages. The host wraps its in-app
/// `HostServer` (fan-out to all connected app peers). Each peer wraps a
/// single outbound WebSocket to the host.
protocol TagTransport: AnyObject {
    var onInbound: ((TagMessage) -> Void)? { get set }
    var onPeerConnected: ((String) -> Void)? { get set }
    var onPeerDisconnected: ((String) -> Void)? { get set }
    func send(_ msg: TagMessage)
    func dispose()
}

/// Host-side: wraps a running [HostServer]. Inbound guest messages are
/// parsed as TagMessages; `send` broadcasts to every connected guest. Owns
/// the server — `dispose()` stops it.
final class HostTagTransport: TagTransport {
    private let server: HostServer
    private var guestToPeer: [GuestId: String] = [:]

    var onInbound: ((TagMessage) -> Void)?
    var onPeerConnected: ((String) -> Void)?
    var onPeerDisconnected: ((String) -> Void)?

    init(server: HostServer) {
        self.server = server
        server.onJoin = { [weak self] g in self?.onPeerConnected?(g.value) }
        server.onLeave = { [weak self] g in
            if let pid = self?.guestToPeer.removeValue(forKey: g) {
                self?.onPeerDisconnected?(pid)
            }
        }
        server.onMessage = { [weak self] g, raw in
            guard let self else { return }
            guard let msg = try? TagMessage.decode(raw) else {
                // Not a Tag peer — almost certainly the browser-tier
                // "Join a game" flow ({"type":"join"}). Tag is BLE-proximity
                // and has no code-join path, so reject it immediately rather
                // than leaving the guest stuck on "Connecting…".
                let err = #"{"type":"error","message":"This host is running Tag. Tag uses Bluetooth proximity and can’t be joined with a code."}"#
                self.server.disconnect(g, sending: err)
                return
            }
            if case let .hello(peerId, _) = msg { self.guestToPeer[g] = peerId }
            self.onInbound?(msg)
            // Echo non-hello traffic back so other peers see it. Host's
            // own TagSession already applied any state change locally.
            if case .hello = msg { /* don't echo hello */ } else { self.server.broadcast(raw) }
        }
    }

    func send(_ msg: TagMessage) { server.broadcast(msg.encode()) }
    func dispose() { server.stop() }
}

/// Peer-side: connects one WebSocket to the host. Outbound goes to the host,
/// which fans out to all peers (including us).
final class PeerTagTransport: NSObject, TagTransport, URLSessionWebSocketDelegate, URLSessionDelegate {
    var onInbound: ((TagMessage) -> Void)?
    var onPeerConnected: ((String) -> Void)?
    var onPeerDisconnected: ((String) -> Void)?

    private var task: URLSessionWebSocketTask?
    private var session: URLSession?

    /// Connects to wss://<host>:<port>/ws derived from the host URL.
    init(serverUrl: URL) {
        super.init()
        var comps = URLComponents(url: serverUrl, resolvingAgainstBaseURL: false)!
        comps.scheme = (comps.scheme == "https") ? "wss" : "ws"
        comps.path = "/ws"
        let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
        self.session = session
        let task = session.webSocketTask(with: comps.url!)
        self.task = task
        task.resume()
        receive()
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(.string(let s)):
                if let m = try? TagMessage.decode(s) { self.onInbound?(m) }
                self.receive()
            case .success(.data(let d)):
                if let s = String(data: d, encoding: .utf8),
                   let m = try? TagMessage.decode(s) { self.onInbound?(m) }
                self.receive()
            case .success:
                self.receive()
            case .failure: break
            }
        }
    }

    func send(_ msg: TagMessage) {
        task?.send(.string(msg.encode())) { _ in }
    }

    func dispose() {
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        session?.invalidateAndCancel()
    }

    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust,
              let cert = HostServer.bundledCert,
              SecTrustSetAnchorCertificates(serverTrust, [cert] as CFArray) == errSecSuccess else {
            completionHandler(.performDefaultHandling, nil); return
        }
        SecTrustSetAnchorCertificatesOnly(serverTrust, true)
        var result: SecTrustResultType = .invalid
        SecTrustEvaluate(serverTrust, &result)
        if result == .unspecified || result == .proceed {
            completionHandler(.useCredential, URLCredential(trust: serverTrust))
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
}

/// Local-only transport. Single-device dev mode — messages echo back into
/// [onInbound] so the engine still sees what it sent.
final class LoopbackTagTransport: TagTransport {
    var onInbound: ((TagMessage) -> Void)?
    var onPeerConnected: ((String) -> Void)?
    var onPeerDisconnected: ((String) -> Void)?
    func send(_ msg: TagMessage) { onInbound?(msg) }
    func dispose() {}
}

File: android/app/src/main/java/com/example/ubapp/games/tag/TagTransport.kt
package com.example.ubapp.games.tag

import com.example.ubapp.social.GuestId
import com.example.ubapp.social.HostServer
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener

/**
 * Bidirectional channel for tag messages. The host wraps its in-app
 * [HostServer]; each peer wraps a single outbound WebSocket to the host.
 */
interface TagTransport {
    var onInbound: ((TagMessage) -> Unit)?
    var onPeerConnected: ((String) -> Unit)?
    var onPeerDisconnected: ((String) -> Unit)?
    fun send(msg: TagMessage)
    fun dispose()
}

/** Host-side: wraps a running [HostServer]. */
class HostTagTransport(private val server: HostServer) : TagTransport {
    private val guestToPeer = HashMap<GuestId, String>()

    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null

    init {
        server.onJoin = { g -> onPeerConnected?.invoke(g.value) }
        server.onLeave = { g ->
            guestToPeer.remove(g)?.let { onPeerDisconnected?.invoke(it) }
        }
        server.onMessage = { g, raw ->
            val msg = runCatching { TagMessage.decode(raw) }.getOrNull()
            if (msg == null) {
                // Not a Tag peer — almost certainly the browser-tier
                // "Join a game" flow ({"type":"join"}). Tag is BLE-proximity
                // and has no code-join path, so reject it immediately rather
                // than leaving the guest stuck on "Connecting…".
                val err = """{"type":"error","message":"This host is running Tag. Tag uses Bluetooth proximity and can’t be joined with a code."}"""
                server.disconnect(g, err)
            } else {
                if (msg is TagMessage.Hello) guestToPeer[g] = msg.peerId
                onInbound?.invoke(msg)
                // Echo non-hello traffic back so other peers see it.
                if (msg !is TagMessage.Hello) server.broadcast(raw)
            }
        }
    }

    override fun send(msg: TagMessage) = server.broadcast(msg.encode())
    override fun dispose() = server.stopServer()
}

/** Peer-side: connects one WebSocket to the host. */
class PeerTagTransport private constructor(serverUrl: String) : TagTransport {
    private val client = OkHttpClient()
    private var ws: WebSocket? = null

    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null

    init {
        val req = Request.Builder().url(serverUrl).build()
        ws = client.newWebSocket(req, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                runCatching { onInbound?.invoke(TagMessage.decode(text)) }
            }
        })
    }

    companion object {
        /** Convert "https://host:port/" to "wss://host:port/ws" and connect. */
        fun connect(serverUrl: String): PeerTagTransport {
            val wsUrl = serverUrl
                .replaceFirst("https://", "wss://")
                .replaceFirst("http://", "ws://")
                .trimEnd('/') + "/ws"
            return PeerTagTransport(wsUrl)
        }
    }

    override fun send(msg: TagMessage) { ws?.send(msg.encode()) }
    override fun dispose() {
        ws?.close(1000, null); ws = null
        client.dispatcher.executorService.shutdown()
    }
}

/** Local-only transport for single-device dev mode. */
class LoopbackTagTransport : TagTransport {
    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null
    override fun send(msg: TagMessage) { onInbound?.invoke(msg) }
    override fun dispose() {}
}
